### PR TITLE
Typings for oauth2orize

### PR DESCRIPTION
**Pull request by heycalmdown**
Please fill in this template.

- [x] Prefer to make your PR against the `types-2.0` branch.
- [x] Test the change in your own code.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes).

If adding a new definition:
- [x] The package does not provide its own types, and you can not add them.
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Run `tsc` without errors.
- [x] Include the required [files](https://github.com/DefinitelyTyped/DefinitelyTyped#create-a-new-package) and header. Base these on the README, *not* on an existing project.

### CODE REVIEW

File: oauth2orize/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,93 @@
+// Type definitions for oauth2orize v1.5.1
+// Project: https://github.com/jaredhanson/oauth2orize/
+// Definitions by: Wonshik Kim <https://github.com/wokim/>, Kei Son <https://github.com/heycalmdown>
+// Definitions: https://github.com/borisyankov/DefinitelyTyped
+
+/// <reference types="node" />
+
+import * as http from "http";
+
+declare module "oauth2orize" {
```
Comment (andy-ms) [inline oauth2orize/index.d.ts]: Don't need this wrapper -- just write the module definition at the top level.

File: oauth2orize/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    //authorization(validate: (clientId: string, redirectURI: string, validated: (err: Error, client: any, redirectURI: string) => void) => void): Function;
+    token(options?: any): (req: http.IncomingMessage, res: http.ServerResponse, next: any) => void;
+    errorHandler(options?: any): (err: Error, req: http.IncomingMessage, res: http.ServerResponse, next: any) => void;
+    serializeClient(fn: (client: any, done: (err: Error | null, id: string) => void) => void): void;
+    serializeClient(client: any, done: (err: Error | null, id: string) => void): void;
+    deserializeClient(fn: (id: string, done: (err: Error | null, client?: any | boolean) => void) => void): void;
+    deserializeClient(obj: any, done: (err: Error | null, client?: any | boolean) => void): void;
+  }
+
+  interface ExchangeFunction extends Function {
```
Comment (andy-ms) [inline oauth2orize/index.d.ts]: Why not just use `type ExchangeFunction = (req: http.ServerRequest, res: http.ServerResponse, next: Function) => void;`?
Also, is there a more specific type `next` could have?

File: oauth2orize/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+  export interface ErrorHandlerOptions {
+    mode?: string;
+  }
+
+  export class OAuth2Server {
+    exchange(fn: ExchangeFunction): OAuth2Server;
+    exchange(type: string, fn: ExchangeFunction): OAuth2Server;
+    // Parses requests to obtain authorization
+    authorize(options: AuthorizeOptions, validate: (clientId: string, redirectURI: string, validated: (err: Error, client: any, redirectURI: string) => void) => void): Function;
```
Comment (andy-ms) [inline oauth2orize/index.d.ts]: Is there a more specific return type this could have?

File: oauth2orize/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    serializeClient(client: any, done: (err: Error | null, id: string) => void): void;
+    deserializeClient(fn: (id: string, done: (err: Error | null, client?: any | boolean) => void) => void): void;
+    deserializeClient(obj: any, done: (err: Error | null, client?: any | boolean) => void): void;
+  }
+
+  interface ExchangeFunction extends Function {
+    (req: http.ServerRequest, res: http.ServerResponse, next: Function): void;
+  }
+
+  export module exchange {
```
Comment (andy-ms) [inline oauth2orize/index.d.ts]: `export namespace exchange`

File: oauth2orize/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    }
+
+    function authorizationCode(options: Options, issue: (client: any, code: string, redirectURI: string, issued: IssuedFunction) => void): ExchangeFunction;
+    function authorizationCode(issue: (client: any, code: string, redirectURI: string, issued: IssuedFunction) => void): ExchangeFunction;
+    function code(options: Options, issue: (client: any, code: string, redirectURI: string, issued: IssuedFunction) => void): ExchangeFunction;
+    function code(issue: (client: any, code: string, redirectURI: string, issued: IssuedFunction) => void): ExchangeFunction;
+
+    function clientCredentials(options: Options, issue: (client: any, scope: string[], issued: IssuedFunction) => void): ClientCredentialFunction;
+    function clientCredentials(options: Options, issue: (client: any, issued: IssuedFunction) => void): ClientCredentialFunction;
+    function clientCredentials(issue: (client: any, scope: string[], issued: IssuedFunction) => void): ClientCredentialFunction;
```
Comment (andy-ms) [inline oauth2orize/index.d.ts]: How does it know what parameters to pass to `issue` at runtime? Does it use `issue.length`?

File: oauth2orize/index.d.ts [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    mode?: string;
+  }
+
+  export class OAuth2Server {
+    exchange(fn: ExchangeFunction): OAuth2Server;
+    exchange(type: string, fn: ExchangeFunction): OAuth2Server;
+    // Parses requests to obtain authorization
+    authorize(options: AuthorizeOptions, validate: (clientId: string, redirectURI: string, validated: (err: Error, client: any, redirectURI: string) => void) => void): Function;
+    authorize(validate: (clientId: string, redirectURI: string, validated: (err: Error | null, client?: any, redirectURI?: string) => void) => void): Function;
+    //authorization(options: AuthorizeOptions, validate: (clientId: string, redirectURI: string, validated: (err: Error, client: any, redirectURI: string) => void) => void): Function;
```
Comment (andy-ms) [inline oauth2orize/index.d.ts]: Why is this left out?

File: oauth2orize/index.d.ts (hunk 5 above)
Comment (heycalmdown) [inline oauth2orize/index.d.ts]: You are right. It checks arity of the passing function.
Anyway, last year when I wrote this definition for personal use it supports up to 3 arity, but now it has 5. I'll update them at next version of definition.

https://github.com/jaredhanson/oauth2orize/blob/0d1f192193591338d018c6be7b967392eed5d482/lib/exchange/clientCredentials.js#L122-L130

File: oauth2orize/index.d.ts (hunk 6 above)
Comment (heycalmdown) [inline oauth2orize/index.d.ts]: it was mistake.

File: oauth2orize/index.d.ts (hunk 2 above)
Comment (andy-ms) [inline oauth2orize/index.d.ts]: Same for `IssuedFunction`

**Pull request merged**